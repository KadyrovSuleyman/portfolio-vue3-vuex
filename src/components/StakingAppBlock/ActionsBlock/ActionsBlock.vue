<script setup lang="ts">

import { computed } from 'vue';
import Div from '@/elements/Div/Div.vue';
import Button from '@/elements/Button/Button.vue';
import Span from '@/elements/Span/Span.vue';
import propsObj from '@/elements/propsObj';
import { useStore } from 'vuex';
import WaitingIcon from './WaitingIcon/WaitingIcon.vue';
import adaptState from './state';
import generateMainButtonProps, { generateUnstakeHandler } from './handlers';

// eslint-disable-next-line no-undef
const props = defineProps({ ...propsObj });
const comp = computed(() => ({ elem: props.elem || 'actionsBlock' }));

const store = useStore();
const state = computed(() => adaptState(store));
const mainButtonProps = computed(() => generateMainButtonProps(adaptState(store)));

const CALLBACK = computed(() => {
  if (!mainButtonProps.value.handler) {
    return () => ({});
  }

  return mainButtonProps.value.handler(store);
});

const unstakeClickHandler = generateUnstakeHandler(store);

</script>

<template>
  <Div :block="props.block" :elem="comp.elem" :mods="props.mods">

    <Button v-if="state.isStaked && !state.isRestakeAvailable"
      :block="comp.elem" :elem="'timeButton'"
    >
      <Span :block="'timeButton'">
        {{ state.restakeCountdown }}
      </Span>
    </Button>

    <Button :block="comp.elem" :elem="'mainButton'"
      :mods="{ waiting: state.isWaiting, disabled: state.disabled }"
      :onClick="CALLBACK" :disabled="state.disabled" v-if="!state.hidden"
    >
      <WaitingIcon
        v-if="state.isWaiting"
        :block="'mainButton'"
      />
      <Span :block="'mainButton'">
        {{ mainButtonProps.text }}
      </Span>
    </Button>

    <Span v-if="state.hidden" :block="comp.elem" :elem="'maxAmountSpan'">
      You have max. amount TKN
    </Span>

    <Button v-if="state.isStaked && state.isRestakeAvailable"
      :block="comp.elem" :elem="'unstakeButton'"
      :onClick="unstakeClickHandler"
    >
      <Span :block="'unstakeButton'">
        Unstake
      </Span>
    </Button>

  </Div>
</template>

<style lang="scss" scoped>
  @use 'styles' as *;
</style>
