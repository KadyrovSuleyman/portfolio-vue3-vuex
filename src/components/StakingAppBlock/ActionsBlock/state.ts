import { Store } from 'vuex';

export interface StateT {
  isWalletConnected: boolean,
  isWalletApproved: boolean,
  isStaked: boolean,
  isReplenishAvailable: boolean,
  isRestakeAvailable: boolean,
  restakeCountdown: string,

  isWaiting: boolean,

  disabled: boolean,

  hidden: boolean,
}

const adaptState = (store: Store<any>): StateT => ({
  isWalletConnected: store.getters['wallet/isConnected'],
  isWalletApproved: store.state.wallet.isApproved,
  isStaked: store.state.stake.isStaked,
  isReplenishAvailable: store.getters['stake/isReplenishAllowed'],
  isRestakeAvailable: store.getters['stake/isRestakeAllowed'],
  restakeCountdown: store.getters['stake/countdown'],

  isWaiting: (
    store.state.waiting.connectWallet
    || store.state.waiting.approveWallet
    || store.state.waiting.stake
    || store.state.waiting.restake
    || store.state.waiting.unstake
  ),

  disabled: !(
    !store.getters['wallet/isConnected']
    || !store.state.wallet.isApproved

    || (
      store.getters['stake/isAllowed']
      && store.getters['stake/isInputValid']
    )

    || store.getters['stake/isReplenishAllowed']

    || store.getters['stake/isRestakeAllowed']
  ),

  hidden: (
    store.state.stake.isStaked
    && (
      !store.getters['stake/isReplenishAllowed']
      && !store.getters['stake/isRestakeAllowed']
    )
  ),
});

export default adaptState;
