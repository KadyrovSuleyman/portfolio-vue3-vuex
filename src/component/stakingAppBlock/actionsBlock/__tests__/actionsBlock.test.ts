import { mount, VueWrapper } from '@vue/test-utils';
import { createStore, Store } from 'vuex';
import ActionsBlock from '../actionsBlock.vue';

jest.mock('../adapter');

let wrapper: VueWrapper<any>;
afterEach(() => {
  wrapper.unmount();
});

let store: Store<any>;
beforeEach(() => {
  store = createStore<any>({
    state: {
      isWalletConnected: false,
      isWalletApproved: false,
      isWaiting: false,
      isStaked: false,
      isReplenishAvailable: false,
      isRestakeAvailable: false,
      restakeCountdown: '00:00:09',
    },
    mutations: {
      change: (state, obj: { [name: string]: boolean | string }) => {
        Object.keys(obj).forEach((index) => {
          state[index] = obj[index];
        });
      },
    },
  });
});

it('waitingIcon renders', () => {
  wrapper = mount(ActionsBlock, {
    global: { plugins: [store] },
  });

  expect(wrapper.find('.actionsBlock').classes()).toEqual(['actionsBlock']);
});

it('watchs props changes', async () => {
  wrapper = mount(ActionsBlock, {
    global: { plugins: [store] },
  });
  expect(wrapper.find('.actionsBlock').classes()).toEqual(['actionsBlock']);

  await wrapper.setProps(
    { mods: { selected: true } },
  );
  expect(wrapper.find('.actionsBlock').classes()).toEqual(['actionsBlock', 'actionsBlock__selected']);

  await wrapper.setProps(
    { mods: { selected: false } },
  );
  expect(wrapper.find('.actionsBlock').classes()).toEqual(['actionsBlock']);
});

it('main button watch the store', async () => {
  wrapper = mount(ActionsBlock, {
    global: { plugins: [store] },
  });
  expect(wrapper.find('.mainButton-span').text()).toBe('Connect wallet');

  store.commit('change', {
    isWaiting: true,
  });
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.mainButton-span').text()).toBe('Waiting');

  store.commit('change', {
    isWaiting: false,
    isWalletConnected: true,
    isWalletApproved: true,
    isStaked: true,
  });
  store.commit('change', {
    isRestakeAvailable: true,
    isReplenishAvailable: false,
  });
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.timeButton-span').exists()).toBeFalsy();
  expect(wrapper.find('.unstakeButton-span').exists()).toBeTruthy();
  expect(wrapper.find('.mainButton-span').text()).toBe('Restake');
  expect(wrapper.find('.unstakeButton-span').text()).toBe('Unstake');

  store.commit('change', {
    isRestakeAvailable: false,
    isReplenishAvailable: true,
  });
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.timeButton-span').exists()).toBeTruthy();
  expect(wrapper.find('.unstakeButton-span').exists()).toBeFalsy();
  expect(wrapper.find('.mainButton-span').text()).toBe('Replenish');
  expect(wrapper.find('.timeButton-span').text()).toBe('00:00:09');
});
