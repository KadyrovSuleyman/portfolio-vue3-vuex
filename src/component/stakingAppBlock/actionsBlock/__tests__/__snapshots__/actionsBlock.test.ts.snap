// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main button watch the store 1`] = `
<div class="actionsBlock">
  <!--v-if--><button class="actionsBlock-mainButton">
    <!--v-if--><span class="mainButton-span">Restake</span>
  </button>
  <!--v-if--><button class="actionsBlock-unstakeButton"><span class="unstakeButton-span"> Unstake </span></button>
</div>
`;

exports[`main button watch the store 2`] = `
<div class="actionsBlock"><button class="actionsBlock-timeButton"><span class="timeButton-span">00:00:09</span></button><button class="actionsBlock-mainButton">
    <!--v-if--><span class="mainButton-span">Replenish</span>
  </button>
  <!--v-if-->
  <!--v-if-->
</div>
`;

exports[`waitingIcon renders 1`] = `
<div class="actionsBlock">
  <!--v-if--><button class="actionsBlock-mainButton">
    <!--v-if--><span class="mainButton-span">Connect wallet</span>
  </button>
  <!--v-if-->
  <!--v-if-->
</div>
`;
